### PR TITLE
Use PDM for lockfiles & env management

### PR DESCRIPTION
**Pull request by adriangb**
This replaces our scripted together requirements management with [PDM](https://pdm.fming.dev/latest/). Our current system has several limitations and can lead to upsies (see https://github.com/pydantic/pydantic/commit/fb2944808b2b216307fccd5babf11c44bfed55eb#diff-3362b72813f83c18d0f1be4e6039d1a1ddc4e16b1324c2a82700d74b990011f9).

I chose PDM because it can work with hatchling as the build backend, which minimizes the number of things that need to change in this PR (vs. something like Poetry which would require a lot more churn).

Selected Reviewer: @samuelcolvin

### CODE REVIEW

File: pydantic/networks.py [hunk 1]
```diff
@@ -325,7 +325,10 @@ def validate_email(value: str) -> tuple[str, str]:
             'value_error', 'value is not a valid email address: {reason}', {'reason': str(e.args[0])}
         ) from e
 
-    return name or parts['local'], parts['email']
+    email = parts.normalized
+    assert email is not None
+    name = name or parts.local_part
+    return name, email
```
Comment (dmontagu) [inline pydantic/networks.py:331]: I'm curious, was this change necessary due to something in this PR? Seems unrelated lol

Comment (dmontagu) [reply pydantic/networks.py:331]: I guess maybe it was a dependency bump thing

File: Makefile [hunk 2]
```diff
@@ -1,31 +1,36 @@
...
-rebuild-lockfiles:       ## Rebuild lockfiles from scratch, updating all dependencies
-	bash requirements/rebuild.sh
+rebuild-lockfiles: .pdm       ## Rebuild lockfiles from scratch, updating all dependencies
+	pdm lock --group :all --dev
 
 .PHONY: format
 format:
-	black $(sources)
-	ruff --fix $(sources)
+	pdm run black $(sources)
```
Comment (dmontagu) [inline Makefile:27]: does this mean you have to do `pdm run <command>` any time you want to run something in your dev environment? I guess maybe you are supposed to activate the environment with pdm or whatever and then it's not necessary?

Does the `--no-isolation` flag used below cause it to not create an environment and just use whatever is active?

Comment (adriangb) [reply Makefile:27]: You can still activate the environment and then just run the thing. But `pdm run` does that for you.

**Comment by dmontagu** [comment]
Okay I looked at the docs page for PDM and it looks pretty well-thought-out. I do think it's a good idea to rework the dependency management stuff, given the "upsies" lol

File: Makefile (hunk 2 above)
Comment (adriangb) [reply Makefile:27]: > Does the --no-isolation flag used below cause it to not create an environment and just use whatever is active?

I believe so. I guess we'll see when that cron job CI file runs.

**Comment by adriangb** [comment]
I'll say that while PDM seems well thought out, it is opinionated and seems to favor correctness over user friendliness. The docs are also not great in my opinion, they tell you an option exists but not how to use it or what it really means. Anyhow, it works.

File: pydantic/networks.py (hunk 1 above)
Comment (adriangb) [reply pydantic/networks.py:331]: Yeah. I accidentally bumped the locked version (because I recreated all lockfiles) and figured I might as well fix this so we can bump it before 2.0 and not have to support multiple versions. But I'm also happy to move that to another PR.

**Comment by adriangb** [comment]
please review

**Comment by adriangb** [comment]
I also think there's something to be said for using a single tool (PDM?) for managing both the build backend, publishing and dependencies/lockfiles.

File: pydantic/networks.py (hunk 1 above)
Comment (dmontagu) [reply pydantic/networks.py:331]: I'm good with fixing it here, easiest to just do it in one go.

**Comment by samuelcolvin** [comment]
In principle I'm good with this - nice to have a solution to the requirements mess.

Can we update the contributions section to explain how to use this.

I'll try this out today.

File: Makefile [hunk 3]
```diff
@@ -1,71 +1,76 @@
 .DEFAULT_GOAL := all
 sources = pydantic tests docs/plugins
 
+.PHONY: .pdm
+.pdm:  ## Check that PDM is isntallced
+	@pdm -V || echo 'Please install PDM: https://pdm.fming.dev/latest/\#installation'
+
+.PHONY: .pre-commit
+.pre-commit:  ## Check that PDM is isntallced
```
Comment (hramezani) [inline Makefile:9]:
```suggestion
.pre-commit:  ## Check that pre-commit is installed
```

File: docs/contributing.md [hunk 4]
```diff
@@ -40,29 +40,27 @@ You'll need to have a version between **Python 3.7 and 3.11**, **virtualenv**, *
...
-source env/bin/activate
-# Building docs requires 3.8. If you don't need to build docs you can use
-# whichever version; 3.7 will work too.
+# 2. Install PDM and pre-commit
+# We use pipx here, for other options see:
+# https://pdm.fming.dev/latest/#installation
+# https://pre-commit.com/#install
+# To get pipx itself:
+# https://pypa.github.io/pipx/
+pipx install pdm pre-commit
```
Comment (samuelcolvin) [inline docs/contributing.md:49]: I think simpler to use `python -m venv` and `pip` here since that's the tool most people will understand and have installed.

File: Makefile [hunk 5]
```diff
@@ -1,71 +1,76 @@
...
-rebuild-lockfiles:       ## Rebuild lockfiles from scratch, updating all dependencies
-	bash requirements/rebuild.sh
+rebuild-lockfiles: .pdm       ## Rebuild lockfiles from scratch, updating all dependencies
+	pdm lock --dev --group :all
 
 .PHONY: format
-format:
-	black $(sources)
-	ruff --fix $(sources)
+format: .pdm
```
Comment (samuelcolvin) [inline Makefile:26]: I don't think we need ` .pdm` on this or later commands - we don't check everything else is installed.

If you do want this behaviour, maybe remove `PHONY` from the `.pdf` recipe and make will check it's been run without calling `pdm`.

File: docs/contributing.md (hunk 4 above)
Comment (samuelcolvin) [reply docs/contributing.md:49]: you could add a comment saying "pipx also works, if that's what you're used to".

**Review by samuelcolvin — COMMENTED** [review]
otherwise I think this is excellent - I've just tried it and pdm looks great.

(Biggest problem is I keep mis-typing it "pdf")

File: docs/contributing.md (hunk 4 above)
Comment (adriangb) [reply docs/contributing.md:49]: The point is you want to install these tools _outside_ of the venv. We could say `pip install pdm pre-commit` but IMO using `pipx` is much nicer than global installs because of the isolation.
